NameLayerMain/src/vg/civcraft/mc/namelayer/gui/MainGroupGUI.java

```diff
@@ -0,0 +1,1272 @@
+			ISUtils.addLore(
+					is,
+					ChatColor.AQUA
+							+ String.valueOf(g.getAllMembers(PlayerType.OWNER)
+									.size()) + " owner");
+		}
+		if (hasGroupStatsPerm) {
+			ISUtils.addLore(
+					is,
+					ChatColor.DARK_AQUA
+							+ String.valueOf(g.getAllMembers().size())
+							+ " total group members");
+			ISUtils.addLore(is, ChatColor.DARK_AQUA + "Group owner: "
+					+ ChatColor.YELLOW + NameAPI.getCurrentName(g.getOwner()));
+		}
+		c = new DecorationStack(is);
+		return c;
+	}
+
+	/**
+	 * Utility to get a "nice" version of the rank the given player has in the
+	 * group this gui is based on
+	 * 
+	 * @param uuid
+	 *            Player whose rank name should be checked
+	 * @return Rankname of the given player ready to be put into the gui or null
+	 *         if the given UUID was null or the player didn't have an explicit
+	 *         rank in the group
+	 */
+	private String getRankName(UUID uuid) {
+		if (uuid == null) {
+			return null;
+		}
+		PlayerType pType = g.getPlayerType(uuid);
+		if (pType == null) {
+			return null;
+		}
+		return getDirectRankName(pType);
+	}
+
+	public static String getDirectRankName(PlayerType pType) {
+		String res = pType.toString().toLowerCase();
+		char[] chars = res.toCharArray();
+		chars[0] = new String(new char[] { chars[0] }).toUpperCase()
+				.toCharArray()[0];
+		res = new String(chars);
+		if (res.endsWith("s")) {
+			return res.substring(0, res.length() - 1);
+		}
+		return res;
+	}
+
+	/**
+	 * Gets the permission needed to make changes to the given PlayerType
+	 * 
+	 * @param pt
+	 *            PlayerType to get permission for
+	 * @return Permission belonging to the given PlayerType or null if the
+	 *         permission was NOT_BLACKLISTED or null
+	 */
+	public static PermissionType getAccordingPermission(PlayerType pt) {
+		switch (pt) {
+		case MEMBERS:
+			return PermissionType.getPermission("MEMBERS");
+		case MODS:
+			return PermissionType.getPermission("MODS");
+		case ADMINS:
+			return PermissionType.getPermission("ADMINS");
+		case OWNER:
+			return PermissionType.getPermission("OWNER");
+		}
+		return null;
+	}
+
+	/**
+	 * Utility to determine whether the player is being promoted or demoted
+	 */
+	private static String demoteOrPromote(PlayerType oldRank,
+			PlayerType newRank, boolean upperCaseFirstLetter) {
+		int oldIndex = 0;
+		int newIndex = 0;
+		for (int i = 0; i < PlayerType.values().length; i++) {
+			if (PlayerType.values()[i] == oldRank) {
+				oldIndex = i;
+			}
+			if (PlayerType.values()[i] == newRank) {
+				newIndex = i;
+			}
+		}
+		String res = oldIndex <= newIndex ? "promote" : "demote";
+		if (upperCaseFirstLetter) {
+			return res.substring(0, 1).toUpperCase() + res.substring(1);
+		}
+		return res;
+	}
+
+	private List<Clickable> getRecursiveInheritedMembers(Group g) {
+		List<Clickable> clicks = new LinkedList<Clickable>();
+		if (g.hasSuperGroup()) {
+			clicks.addAll(getRecursiveInheritedMembers(g.getSuperGroup()));
```

Comment:
Ah, so this here is susceptible to cyclic linking as well; maybe we should just add a startup check that scans & removes any cycles?